// wdio-sauce-service/build/sauce-service.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

var jobDataProperties = ['name', 'tags', 'public', 'build', 'custom-data'];

var SauceService = (function () {
    function SauceService() {
        _classCallCheck(this, SauceService);
    }

    _createClass(SauceService, [{
        key: 'before',

        /**
         * gather information about runner
         */
        value: function before(capabilities) {
            this.sessionId = global.browser.sessionId;
            this.capabilities = capabilities;
            this.auth = global.browser.requestHandler.auth || {};
            this.sauceUser = this.auth.user;
            this.sauceKey = this.auth.pass;
            this.sauceAPIUri = 'https://saucelabs.com/rest/v1/' + this.sauceUser + '/jobs/' + this.sessionId;
        }
    }, {
        key: 'beforeSuite',
        value: function beforeSuite(suite) {
            this.suiteTitle = suite.title;
        }
    }, {
        key: 'beforeTest',
        value: function beforeTest(test) {
            if (!this.sauceUser || !this.sauceKey) {
                return;
            }

            global.browser.execute('sauce:context=' + test.parent + ' - ' + test.title);
        }
    }, {
        key: 'beforeFeature',
        value: function beforeFeature(feature) {
            if (!this.sauceUser || !this.sauceKey) {
                return;
            }

            this.suiteTitle = feature.getName();
            global.browser.execute('sauce:context=Feature: ' + this.suiteTitle);
        }
    }, {
        key: 'beforeScenario',
        value: function beforeScenario(scenario) {
            if (!this.sauceUser || !this.sauceKey) {
                return;
            }

            global.browser.execute('sauce:context=Scenario: ' + scenario.getName());
        }

        /**
         * update Sauce Labs job
         */
    }, {
        key: 'after',
        value: function after(failures) {
            var _this = this;

            if (!this.sauceUser || !this.sauceKey) {
                return;
            }

            return new Promise(function (resolve, reject) {
                return _request2['default'].put(_this.sauceAPIUri, {
                    json: true,
                    auth: {
                        user: _this.sauceUser,
                        pass: _this.sauceKey
                    },
                    body: _this.getBody(failures)
                }, function (e, res, body) {
                    if (e) {
                        return reject(e);
                    }
                    global.browser.jobData = body;
                    resolve(body);
                });
            });
        }

        /**
         * massage data
         */
    }, {
        key: 'getBody',
        value: function getBody(failures) {
            var body = {};

            /**
             * set default values
             */
            body.name = this.suiteTitle;

            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = jobDataProperties[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var prop = _step.value;

                    if (!this.capabilities[prop]) {
                        continue;
                    }

                    body[prop] = this.capabilities[prop];
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator['return']) {
                        _iterator['return']();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            body.passed = failures === 0;
            return body;
        }
    }]);

    return SauceService;
})();

exports['default'] = SauceService;
module.exports = exports['default'];
